pex/third_party/__init__.py

```diff
@@ -0,0 +1,391 @@
+    :param str prefix: The import prefix the installed importer will be responsible for.
+    :param path_items: The paths relative to ``root`` containing modules to expose for import under
+                       ``prefix``.
+    :param str root: The root path of the distribution containing the vendored code. NB: This is the
+                     the path to the pex code, which serves as the root under which code is vendored
+                     at ``pex/vendor/_vendored``.
+    :param str warning: An optional warning to emit if any imports are made through the installed
+                        importer.
+    :return:
+    """
+    root = cls._abs_root(root)
+    importables = tuple(cls._iter_importables(root=root, path_items=path_items, prefix=prefix))
+    vendor_importer = cls(root=root,
+                          importables=importables,
+                          uninstallable=uninstallable,
+                          warning=warning)
+    if warning:
+      warnings.filterwarnings('default', category=DeprecationWarning, module=__name__)
+    sys.meta_path.insert(0, vendor_importer)
+    _tracer().log('Installed {}'.format(vendor_importer), V=3)
+    return vendor_importer
+
+  @classmethod
+  def uninstall_all(cls):
+    """Uninstall all uninstallable VendorImporters and uninmport the modules they loaded."""
+    for vendor_importer in cls._iter_all_installed_vendor_importers():
+      vendor_importer.uninstall()
+
+  def __init__(self, root, importables, uninstallable=True, warning=None):
+    self._root = root
+    self._importables = importables
+
+    self._uninstallable = uninstallable
+    self._warning = warning
+
+    self._loaders = []
+
+  def uninstall(self):
+    """Uninstall this importer if possible and un-import any modules imported by it."""
+    if not self._uninstallable:
+      _tracer().log('Not uninstalling {}'.format(self), V=9)
+      return
+
+    if self in sys.meta_path:
+      sys.meta_path.remove(self)
+      maybe_exposed = frozenset(os.path.join(self._root, importable.path)
+                                for importable in self._importables)
+      sys.path[:] = [path_item for path_item in sys.path if path_item not in maybe_exposed]
+      for loader in self._loaders:
+        loader.unload()
+      _tracer().log('Uninstalled {}'.format(self), V=3)
+
+  # The PEP-302 finder API.
+  # See: https://www.python.org/dev/peps/pep-0302/#specification-part-1-the-importer-protocol
+  def find_module(self, fullname, path=None):
+    for importable in self._importables:
+      loader = importable.loader_for(fullname)
+      if loader is not None:
+        self._loaders.append(loader)
+        if self._warning:
+          warnings.warn('Found loader for `import {}`:\n\t{}'.format(fullname, self._warning),
+                        category=DeprecationWarning)
+        return loader
+    return None
+
+  def _expose(self, paths):
+    for importable in self._importables:
+      if importable.path in paths:
```

Comment:
Should it be an error to request to expose something that isn't in the importables?